### PR TITLE
feat: add `expectedSlippage` to the response of `/swap/v1/price` endpoint

### PR DESCRIPTION
**Pull request by novaliu86**
# Description

Add a new class `SlippageModelManager` to periodically read the slippage model data file in S3. There will be one such file per `chainId` and the files are created and updated by data team with an automatic daily job.

The `SlippageModelManager` checks for update every 1 minute with a `headObject` method which only reads the metadata of the S3 file. `SlippageModelManager` uses the `LastModified` timestamp to determine whether a refresh is necessary, and if yes it will read the full content of the S3 file. `SlippageModelManager` also cache the slippage model data in memory so that the consumer of it, `SwapService`, always reads from memory.

The new optional flag `slippageModel` on the integrator's config is used to determine whether to send the slippage data to an integrator. As the first step we will only enable it for Matcha. If enabled, two fields will be added to the responses of `/price` calls: `expectedSlippage`, and `expectedBuyAmount` if `sellAmount` is given, or `expectedSellAmount` if `buyAmount` is given.

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

**Review by dekz — COMMENTED** [review]
Ensuring it fail safes looks pretty good, I can't see any immediate cases where an S3 outage or a bad schema would result in a failure to quote.

File: src/services/swap_service.ts [hunk 1]
```diff
@@ -475,6 +477,29 @@ export class SwapService {
...
+                apiSwapQuote.expectedSlippage = this._slippageModelManager.calculateExpectedSlippage(
+                    buyToken,
+                    sellToken,
+                    apiSwapQuote.buyAmount,
+                    apiSwapQuote.sellAmount,
+                    slippagePercentage ?? 0,
+                    apiSwapQuote.sources,
+                );
+            } else {
+                apiSwapQuote.expectedSlippage = new BigNumber(0);
```
[the remark this answers is not in the document]
Comment (novaliu86) [reply src/services/swap_service.ts:491]: Sorry I did not explicitly call out that the `expectedSlippage` is actually a weighted average. Just added some comment to the function `SlippageModelManager.calculateExpectedSlippage()`

File: src/services/swap_service.ts [hunk 2]
```diff
@@ -475,6 +477,29 @@ export class SwapService {
...
+            }
+
+            if (marketSide === MarketOperation.Sell) {
+                apiSwapQuote.expectedBuyAmount = apiSwapQuote.buyAmount.times(apiSwapQuote.expectedSlippage.plus(1));
+            } else {
+                apiSwapQuote.expectedSellAmount = apiSwapQuote.sellAmount.times(
```
[the remark this answers is not in the document]
Comment (novaliu86) [reply src/services/swap_service.ts:497]: I see... The purpose of these two values are for front end to easily compare with price from RFQm. If it is a buy the price is represented by sellAmount as buyAmount is given. If it is a sell the price is represented by buyAmount as sellAmount is given.
For a buy operation, both swap and RFQm would return a sellAmount for the given buyAmount, and Matcha will simply compare RFQm `sellAmount` with the `expectedSellAmount` from swap API with slippage accounted.

**Comment by novaliu86** [comment]
> Ensuring it fail safes looks pretty good, I can't see any immediate cases where an S3 outage or a bad schema would result in a failure to quote.

Thank you for your careful review! I made two changes according to your suggestions, but reverted both after thinking more about them. See my reasoning under your comments. Feel free to let me know if you don't agree or have other suggestions :)

File: src/utils/slippage_model_manager.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    public calculateExpectedSlippage(
+        buyToken: string,
+        sellToken: string,
+        buyAmount: BigNumber,
+        sellAmount: BigNumber,
+        maxSlippageRate: number,
+        sources: GetSwapQuoteResponseLiquiditySource[],
+    ): BigNumber {
+        const maxSlippageInBps = new BigNumber(maxSlippageRate * ONE_IN_BASE_POINTS);
+        let expectedSlippage: BigNumber = new BigNumber(0);
```
[the remark this answers is not in the document]
Comment (novaliu86) [reply src/utils/slippage_model_manager.ts:127]: The returned `expectedSlippage` is a weighted average of `expectedSlippage` of all sources. We probably don't want to return a `null` if `slippageModel` for any source is missing...

**Pull request merged**